src/twisted/web/http.py

```diff
@@ -1825,75 +1823,97 @@ class _ChunkedTransferDecoder:
-            return rest
+    def _dataReceived_CHUNK_LENGTH(self) -> bool:
+        """
+        Read the chunk size line, ignoring any extensions.
+
+        @returns: C{True} once the line has been read and removed from
+            C{self._buffer}.  C{False} when more data is required.
+
+        @raises _MalformedChunkedDataError: when the chunk size cannot be
+            decoded.
+        """
+        eolIndex = self._buffer.find(b"\r\n", self._start)
+        if eolIndex == -1:
+            self._start = len(self._buffer) - 1
+            return False
+
+        endOfLengthIndex = self._buffer.find(b";", 0, eolIndex)
+        if endOfLengthIndex == -1:
+            endOfLengthIndex = eolIndex
+        try:
+            length = int(self._buffer[0:endOfLengthIndex], 16)
+        except ValueError:
+            raise _MalformedChunkedDataError("Chunk-size must be an integer.")
+
+        if length < 0:
+            raise _MalformedChunkedDataError("Chunk-size must not be negative.")
+        elif length == 0:
+            self.state = "TRAILER"
         else:
-            self._buffer = data
-            return b""
+            self.state = "BODY"
+
+        self.length = length
+        del self._buffer[0 : eolIndex + 2]
+        self._start = 0
+        return True
 
-    def _dataReceived_CRLF(self, data):
-        if data.startswith(b"\r\n"):
+    def _dataReceived_CRLF(self) -> bool:
+        if self._buffer.startswith(b"\r\n"):
             self.state = "CHUNK_LENGTH"
-            return data[2:]
+            del self._buffer[0:2]
+            return True
         else:
-            self._buffer = data
-            return b""
+            return False
 
-    def _dataReceived_TRAILER(self, data):
-        if data.startswith(b"\r\n"):
-            data = data[2:]
+    def _dataReceived_TRAILER(self) -> bool:
+        if self._buffer.startswith(b"\r\n"):
+            data = memoryview(self._buffer)[2:].tobytes()
+            del self._buffer[:]
             self.state = "FINISHED"
             self.finishCallback(data)
-        else:
-            self._buffer = data
-        return b""
+        return False
 
-    def _dataReceived_BODY(self, data):
-        if len(data) >= self.length:
-            chunk, data = data[: self.length], data[self.length :]
-            self.dataCallback(chunk)
+    def _dataReceived_BODY(self) -> bool:
+        if len(self._buffer) >= self.length:
+            chunk = memoryview(self._buffer)[: self.length].tobytes()
+            del self._buffer[: self.length]
             self.state = "CRLF"
-            return data
-        elif len(data) < self.length:
-            self.length -= len(data)
-            self.dataCallback(data)
-            return b""
+            self.dataCallback(chunk)
+        elif self._buffer:
+            chunk = bytes(self._buffer)
+            self.length -= len(chunk)
+            self._buffer = bytearray()
+            self.dataCallback(chunk)
+        return True
```

Comment:
Just a minor style suggestion.

I think that this refactoring will also solve the missing branch coverage that is preventing this diff for being 100%

In Codecov, yellow is missing branch coverage.

https://app.codecov.io/gh/twisted/twisted/compare/1551/diff#D1L1893

```suggestion
            self.dataCallback(chunk)
            return True

        if not self._buffer:
             # No data received yet for the current chunk.
            return True

        # We have data and we are not at the end of the chunk.
        chunk = bytes(self._buffer)
        self.length -= len(chunk)
        self._buffer = bytearray()
        self.dataCallback(chunk)
        return True
```